### ISSUE TITLE
4.1 doesn't work with Typescript

### ISSUE THREAD

**Issue by Hotell** [skatejs/skatejs#936]
The new implementation of API on prototype doesn't  work with Typescript which is very unfortunate.

Problem is probably in Typescript implementation of `extends`.

It works only with old skate js API's ( define and definition via object )

Short description of possible issue:

- problem starts here ( within `connectedCallback` ): https://github.com/skatejs/skatejs/blob/master/src/api/component.js#L186

- which calls `syncPropsToAttrs(this)`, where `this` is the custom element constructor. Problem is that there is no `props` static property  on it so this call (https://github.com/skatejs/skatejs/blob/master/src/api/component.js#L34) will return `undefined` and when you call `undefined`.forEach well it throws...

![skate-js-4dot1-ts-bug](https://cloud.githubusercontent.com/assets/1223799/20556827/1e867b3c-b169-11e6-8eae-7f8badd19ec8.gif)

Here is the repo which reproduces the error:
https://github.com/Hotell/skate-and-raw-WC

and here is skate component https://github.com/Hotell/skate-and-raw-WC/blob/master/src/app/skate-wc/yo-smile/YoSmile.tsx

just hit `yarn` || `npm i` to install dependencies, then `yarn start` || `npm start` and open devTools, to debug it.

**Comment by vvakame** [comment]
related issue https://github.com/Microsoft/TypeScript/issues/3610

**Comment by Hotell** [comment]
also just one more note, older versions of skate did work with Typescript

**Comment by vascofernandes** [comment]
Would be nice to have this fixed.
I always go TS first these days.

**Comment by andreawyss** [comment]
@Hotell please check it PR #932 (Cache propConfigs returned from static get props)
fixes this issue that you are having with SkateJs 4.1 and TypeScript 
If it does than have @treshugart review my changes.
Thank You

**Comment by vvakame** [comment]
I made work around. check [this change](https://github.com/vvakame/til/commit/6d36167769e87d5235ea6f55fd01f1258936db53).

**Comment by Hotell** [comment]
nice @vvakame! , 

you don't need to import tslib manually, just hit `"importHelpers": true,` in tsconfig.json and just override the global._extends as you do already

**Comment by vvakame** [comment]
@Hotell I tried it now. but it doesn't work as expected.
but following code generated.
```
	"use strict";
	var tslib_1 = __webpack_require__(1);
	__webpack_require__(2);
	var skate = __webpack_require__(3);
	window.__extends = function (d, b) {
	    Object.setPrototypeOf(d, b);
	    var __ = function () { this.constructor = d; };
	    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
	};
	var CountUpComponent = (function (_super) {
	    tslib_1.__extends(CountUpComponent, _super);
	    function CountUpComponent() {
	        return _super.apply(this, arguments) || this;
	    }
```

**Comment by Hotell** [comment]
yeah doesn't work for me as well :( so we have to manually import `tslib` and then override `__extends`

**Comment by andreawyss** [comment]
@Hotell I would not change the prototype of your derived class to be the prototype of your base class. You need to maintain the correct prototype chain in place.
Try to use this code that uses getOwnPropertyNames:

`
var __extends = (this && this.__extends) || function (d, b) {

    Object.getOwnPropertyNames(b).forEach(function(p) {
      d[p] = b[p];
    });

    function __() {
      this.constructor = d;
    }

    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
  };

`

**Comment by Hotell** [comment]
that doesn't work @andreawyss 

for now I'm rather overriding typescript `__extends` with babel's `_inherits`.

Works like a charm, although yeah I'm not big fan of using `setPrototypeOf` and not just for perf reasons...

```typescript
function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }
  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

(window as any).__extends = _inherits;
```

**Comment by andreawyss** [comment]
Yes, I like it.

**Comment by Hotell** [comment]
so I can add to docs how to skate 4.1 work with TS, WDYT @treshugart ?, 

because I don't think that this will be fixed anytime soon within TS, because it would be a huge Breaking Change

**Comment by Hotell** [comment]
@andreawyss 

> When you tried my version did you check the one that skips over 'length', 'name', 'prototype' ?

Nope, but anyway, that's to much nitpicking magic for my taste, I think by using Babel's inherits method, the problem is solved at this moment. WDYT?

If @treshugart gives me go, I'll submit PR to docs

**Comment by andreawyss** [comment]
@Hotell I think you should give it a try.
It follows the original pattern of TypeScript __extends. 
The TypeScript team can advise us on which solution they recommend us to use.

**Comment by treshugart** [comment]
@Hotell go for it. I think it's at least a good starting point for now, that matches the current state of affairs. Hopefully this will be addressed soon. This feels like a refactor to TS would solve this? cc @andreawyss

**Comment by andreawyss** [comment]
Yes, TS is still incorrectly generating static class member properties enumerable=true.
But at least then a class generated with TS can then be extended with ES6+Babel code.

**Comment by Hotell** [comment]
yup (breaking)change on TS side is needed

**Comment by DanielRosenwasser** [comment]
We'll be discussing this in the near future. This might be something to keep an eye out for in the 2.2 timeframe.

**Comment by treshugart** [comment]
Going to close this as it seems we've gotten as far as we can. As per #972 we're going to try Flow as well and see how that works out. We can always switch later after we've had experience trying both.

**Comment by vvakame** [comment]
FYI: https://github.com/Microsoft/TypeScript/pull/12488 this issue will be solved in TypeScript 2.2
